File: src/leetCode/Ex114.java
package leetCode;

public class Ex114 {
    public void flatten(TreeNode root) {
        helper(root);
    }

    private TreeNode helper(TreeNode root){
        if(root==null){
            return null;
        }
        var left = root.left;
        var right = root.right;

        var pre = helper(left);
        var preRight = helper(right);
        if(pre !=null){
            pre.right = right;
            root.right = left;
            root.left = null;
        }
        if(preRight==null){
            if(pre==null){
                return root;
            }else{
                return pre;
            }
        }
        return preRight;
    }
}

File: src/leetCode/ex114.kt
package leetCode

import java.util.*
import kotlin.collections.HashSet

fun flatten(root: TreeNode?): Unit {
    if (root==null){
        return
    }
    var head :TreeNode?= null
    val stack = Stack<TreeNode>()
    val set = HashSet<TreeNode>()
    stack.push(root)
    while (stack.isNotEmpty()){
        val t = stack.peek()
        if (set.contains(t)){
            stack.pop()
            t.left=null
            t.right=head
            head = t
        }else{
            set.add(t)
            if (t.left!=null){
                stack.push(t.left)
            }
            if (t.right!=null){
                stack.push(t.right)
            }
        }
    }
}